Source/Adobe.Target.Client/OnDevice/DecisioningRuleExecutor.cs

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2021 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+namespace Adobe.Target.Client.OnDevice
+{
+    using System.Collections.Generic;
+    using System.Linq;
+    using Adobe.Target.Client.Extension;
+    using Adobe.Target.Client.Model.OnDevice;
+    using Adobe.Target.Client.OnDevice.Collator;
+    using Adobe.Target.Client.Service;
+    using Adobe.Target.Client.Util;
+    using Adobe.Target.Delivery.Model;
+    using JsonLogic.Net;
+    using Newtonsoft.Json.Linq;
+
+    internal sealed class DecisioningRuleExecutor
+    {
+        private const string Allocation = "allocation";
+        private const string State = "state";
+        private const string ResponseTokenDecisioningMethod = "activity.decisioningMethod";
+        private const string OnDevice = "on-device";
+
+        private readonly TargetClientConfig clientConfig;
+        private readonly JsonLogicEvaluator evaluator;
+
+        public DecisioningRuleExecutor(TargetClientConfig clientConfig)
+        {
+            this.clientConfig = clientConfig;
+            this.evaluator = new JsonLogicEvaluator(EvaluateOperators.Default);
+        }
+
+        public IDictionary<string, object> ExecuteRule(
+            IDictionary<string, object> localContext,
+            RequestDetailsUnion details,
+            string visitorId,
+            OnDeviceDecisioningRule rule,
+            ISet<string> responseTokens)
+        {
+            localContext.Remove(Allocation);
+            localContext.Add(Allocation, this.ComputeAllocation(visitorId, rule));
+            var matched = this.EvaluateRule((JObject)rule.Condition, localContext.ToExpandoObject());
+            if (!matched)
+            {
+                return null;
+            }
+
+            return this.ReplaceCampaignMacros(rule, this.GetConsequence(responseTokens, rule, localContext), details);
+        }
+
+        private static IDictionary<string, object> GetGeoResponseTokens(IDictionary<string, object> localContext, ISet<string> responseTokenKeys)
+        {
+            var result = new Dictionary<string, object>();
+            if (!localContext.TryGetValue(OnDeviceDecisioningService.ContextKeyGeo, out var geoContext))
+            {
+                return result;
+            }
+
+            foreach (var geo in (Dictionary<string, object>)geoContext)
+            {
```

Comment:
Could we add proper method and extract the ternary

Reply:
Fixed